src/main/java/org/tugraz/sysds/runtime/lineage/LineageCache.java

```diff
@@ -1,60 +1,307 @@
+		boolean sparse = MatrixBlock.evalSparseFormatInMemory(r, c, nnz);
+		
+		double nflops = 0;
+		CPType cptype = CPInstructionParser.String2CPInstructionType.get(inst.getOpcode());
+		//TODO: All other relevant instruction types.
+		switch (cptype)
+		{
+			case MMTSJ:
+				MMTSJType type = ((MMTSJCPInstruction)inst).getMMTSJType();
+				if (type.isLeft())
+					nflops = !sparse ? (r * c * s * c /2):(r * c * s * c * s /2);
+				else
+					nflops = !sparse ? ((double)r * c * r/2):(r*c*s + r*c*s*c*s /2);
+				break;
+				
+			default:
+				throw new DMLRuntimeException("Lineage Cache: unsupported instruction: "+inst.getOpcode());
+		}
+		return nflops / (2L * 1024 * 1024 * 1024);
+	}
+
+	// ---------------- I/O METHODS TO LOCAL FS -----------------
+	
+	private static void spillToLocalFS() {
+		long t0 = DMLScript.STATISTICS ? System.nanoTime() : 0;
+		if (outdir == null) {
+			outdir = LocalFileUtils.getUniqueWorkingDir(LocalFileUtils.CATEGORY_LINEAGE);
+			LocalFileUtils.createLocalFileIfNotExist(outdir);
+		}
+		String outfile = outdir+"/"+_cache.get(_end._key)._key.getId();
+		try {
+			LocalFileUtils.writeMatrixBlockToLocal(outfile, _cache.get(_end._key)._value);
+		} catch (IOException e) {
+			throw new DMLRuntimeException ("Write to " + outfile + " failed.", e);
+		}
+		if (DMLScript.STATISTICS) {
+			long t1 = System.nanoTime();
+			LineageCacheStatistics.incrementFSWriteTime(t1-t0);
+			LineageCacheStatistics.incrementFSWrites();
+		}
+
+		_spillList.put(_end._key, outfile);
+	}
+	
+	private static MatrixBlock readFromLocalFS(Instruction inst, LineageItem key) {
+		long t0 = DMLScript.STATISTICS ? System.nanoTime() : 0;
+		MatrixBlock mb = null;
+		// Read from local FS
+		try {
+			mb = LocalFileUtils.readMatrixBlockFromLocal(_spillList.get(key));
+		} catch (IOException e) {
+			throw new DMLRuntimeException ("Read from " + _spillList.get(key) + " failed.", e);
+		}
+		// Restore to cache
+		LocalFileUtils.deleteFileIfExists(_spillList.get(key), true);
+		_spillList.remove(key);
+		put(inst, key, mb);
+		if (DMLScript.STATISTICS) {
+			long t1 = System.nanoTime();
+			LineageCacheStatistics.incrementFSReadTime(t1-t0);
+			LineageCacheStatistics.incrementFSHits();
+		}
+		return mb;
+	}
+
+	//------------------ LINKEDLIST MAINTENANCE METHODS -------------------
+	
+	public static void delete(Entry entry) {
+		if (entry._prev != null)
+			entry._prev._next = entry._next;
+		else
+			_head = entry._next;
+		if (entry._next != null)
+			entry._next._prev = entry._prev;
+		else
+			_end = entry._prev;
+	}
+	
+	public static void setHead(Entry entry) {
+		entry._next = _head;
+		entry._prev = null;
+		if (_head != null)
+			_head._prev = entry;
+		_head = entry;
+		if (_end == null)
+			_end = _head;
+	}
+
+	private static void removeEntry(double space) {
+		if (DMLScript.STATISTICS)
+			_removelist.add(_end._key);
+		_cache.remove(_end._key);
+		_cachesize -= space;
+		delete(_end);
+	}
+	
+	private static class Entry {
+		LineageItem _key;
```

Comment:
It would be great if we could remove this class as there are way too many `Entry` implementations already.